### PR TITLE
[BIOMAGE-1222] Implement the marker heatmap page in plots and tables

### PR DESCRIPTION
**Pull request by StefanBabukov**
# Background
#### Link to issue 
https://biomage.atlassian.net/secure/RapidBoard.jspa?rapidView=1&projectKey=BIOMAGE&modal=detail&selectedIssue=BIOMAGE-1222
#### Link to staging deployment URL 
https://ui-stefan-ui429-api198.scp-staging.biomage.net/
#### Links to any Pull Requests related to this
https://github.com/biomage-ltd/api/pull/198
#### Anything else the reviewers should know about the changes here

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [ ] Unit tests written
- [ ] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [ ] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+    dispatch(loadCellSets(experimentId));
+  }, []);
+
+  useEffect(() => {
+    if (louvainClustersResolution && config && !config.selectedGenes.length) {
+      dispatch(loadMarkerGenes(experimentId, louvainClustersResolution, plotUuid, config.numGenes));
+    }
+  }, [louvainClustersResolution, hierarchy, config?.numGenes]);
+
+  const sortGenes = (newGenes) => {
```
Comment (ivababukova) [inline src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx:65]: can you add unit tests to check for whether this sorting works, also to verify that we are not sorting when we are removing a gene and any other case you can come up with?

File: src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+const { Panel } = Collapse;
+
+const plotUuid = 'markerHeatmapPlotMain';
+const plotType = 'markerHeatmap';
+
+const route = {
+  path: 'marker-heatmap',
+  breadcrumbName: 'Marker-Heatmap',
+};
+const HeatmapPlot = ({ experimentId }) => {
```
Comment (ivababukova) [inline src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx]: I would call this component `MarkerHeatmap`, just to distinguish it from the heatmap component easier

File: src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+  const dispatch = useDispatch();
+  const config = useSelector((state) => state.componentConfig[plotUuid]?.config);
+  const { expression: expressionData } = useSelector((state) => state.genes);
+  const { error, loading } = expressionData;
+  const cellSets = useSelector((state) => state.cellSets);
+  const { hierarchy, properties } = cellSets;
+  const selectedGenes = useSelector((state) => state.genes.expression.views[plotUuid]?.data) || [];
+  const [currentSearchedGenes, setCurrentSearchedGenes] = useState([]);
```
Comment (ivababukova) [inline src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx]: I can see that the heatmap component in plots and tables doesn't need to have a state variable for the selected genes, but uses `selectedGenes` directly. Why in the case of maker heatmap we need to keep track of this and in the case of the heatmap plot we don't?

File: src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+              onReset={onReset}
+            />
+            <PlotStyling formConfig={plotStylingControlsConfig} config={config} onUpdate={updatePlotWithChanges} defaultActiveKey={['5']} />
+          </Space>
+        </Col>
+      </Row>
+    </div>
+  );
+};
+
```
Comment (ivababukova) [inline src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx]: lots of this and other parts of this file are identical to what we have for the Heatmap component in Plots and Tables -- have you considered reusing the same heatmap component for them and building this MarkerHeatmap component based on it?

File: src/redux/actionTypes/genes.js [hunk 5]
```diff
@@ -60,9 +60,11 @@ const MARKER_GENES_LOADED = `${GENES}/markerGenesLoaded`;
  */
 const MARKER_GENES_ERROR = `${GENES}/markerGenesError`;
 
+const GENE_ORDER_SET = `${GENES}/geneOrderSet`;
```
Comment (ivababukova) [inline src/redux/actionTypes/genes.js]: as I suggested during our call, will be best to keep this elsewhere, since it is something specifically applicable to the marker heatmap in plots and tables, it is not used anywhere else.

File: src/redux/reducers/componentConfig/initialState.js [hunk 6]
```diff
@@ -99,6 +99,41 @@ const heatmapInitialConfig = {
   truncatedValues: false,
 };
 
+// PLOTS & TABLES - Marker heatmap
+const markerHeatmapInitialConfig = {
```
Comment (ivababukova) [inline src/redux/reducers/componentConfig/initialState.js:106]: most of this config is the same as the config of `heatmapInitialConfig`, I would recommend that you modify the code here to copy whatever is in `heatmapInitialConfig` and then just build on top of it. so have something like:

```
const markerHeatmapInitialConfig = {
    ...heatmapInitialConfig,
    groupedTracks: ['louvain'],
    and so on
}
```

File: src/redux/reducers/genes/index.js [hunk 7]
```diff
@@ -3,7 +3,7 @@ import {
   GENES_PROPERTIES_LOADING, GENES_PROPERTIES_LOADED_PAGINATED, GENES_PROPERTIES_ERROR,
   GENES_SELECT, GENES_DESELECT,
   GENES_EXPRESSION_LOADING, GENES_EXPRESSION_LOADED, GENES_EXPRESSION_ERROR,
-  MARKER_GENES_LOADING, MARKER_GENES_LOADED, MARKER_GENES_ERROR,
+  MARKER_GENES_LOADING, MARKER_GENES_LOADED, MARKER_GENES_ERROR, GENE_ORDER_SET,
```
Comment (ivababukova) [inline src/redux/reducers/genes/index.js]: also, this sounds like it is better to be named to something `MARKER_GENE_ORDER_SET` and it definitely should be made clear in the code that it is for plots and tables. This doesn't look like the right place in the store for it

**Comment by vickymorrison** [comment]
Sorry Stefan - I can't review this from the user's perspective because every time I change something it becomes unresponsive.
https://user-images.githubusercontent.com/74908153/129191107-a1dd8e4a-3c65-49e7-8f9c-bbdb30bb7081.mov

**Comment by vickymorrison** [comment]
Does it re-run the 'findmarkers' for the custom cell sets when this option is selected? Or is it still showing markers for the Louvain clusters? Somehow I don't believe the marker genes are being calculated for the custom cells sets as this doesn't look like a marker heatmap for custom cell sets A and B:
Screenshot 2021-08-12 at 14 19 23

File: src/__test__/components/plots-and-tables/markerHeatmap.test.jsx [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+  });
+
+  // marker heatmap renders
+  it('loads initially', async () => {
+    await renderHeatmapPage();
+
+    expect(loadMarkersSpy).toHaveBeenCalled();
+  });
+
+  it('loads marker genes on selecting', async () => {
```
Comment (ivababukova) [inline src/__test__/components/plots-and-tables/markerHeatmap.test.jsx:209]: we need more tests apart from loading marker genes on selecting. 
Tests like "the plot changes on change of a given schema parameter", checking that the schema is updated correctly, checking that the relevant update calls were triggered.
Tests about how would the marker heatmap behave if a given part of the redux store is not defined: what if genes were missing? What if experimentSettings were undefined, etc. 
Try to think about what can possibly go wrong and add tests to check about it.

File: src/redux/actions/genes/loadMarkerGenes.js [hunk 9]
```diff
@@ -4,18 +4,19 @@ import {
 
 import { fetchCachedWork } from '../../../utils/cacheRequest';
 
-const loadMarkerGenes = (experimentId, resolution) => async (dispatch, getState) => {
+const loadMarkerGenes = (
+  experimentId, resolution, plotUuid, numGenes = 5,
```
Comment (ivababukova) [inline src/redux/actions/genes/loadMarkerGenes.js:8]: I think in here you want to be passing `plotType`, not `plotUuid`, because eventually we will have multiple plots of the same type, and then you would want to pass the type, not the uuid which will be unique.

File: src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+    if (!louvainClustersResolution) {
+      dispatch(loadProcessingSettings(experimentId));
+    }
+    dispatch(loadPlotConfig(experimentId, plotUuid, plotType));
+    dispatch(loadCellSets(experimentId));
+  }, []);
+
+  useEffect(() => {
+    if (louvainClustersResolution && config && !config.selectedGenes.length) {
+      dispatch(loadMarkerGenes(experimentId, louvainClustersResolution, plotUuid, config.numGenes));
```
Comment (ivababukova) [inline src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx:62]: better pass plotType, not plotUuid

File: src/__test__/redux/actions/genes/__snapshots__/loadMarkerGenes.test.js.snap [hunk 11]
```diff
@@ -38,6 +38,7 @@ Object {
     "genes": Array [
       "geneA",
     ],
+    "plotUuid": "interactiveHeatmap",
```
Comment (ivababukova) [inline src/__test__/redux/actions/genes/__snapshots__/loadMarkerGenes.test.js.snap:41]: this is plotType, not plotUuid

File: src/redux/actions/genes/index.js [hunk 12]
```diff
@@ -2,10 +2,12 @@ import loadGeneExpression from './loadGeneExpression';
...
 import loadPaginatedGeneProperties from './loadPaginatedGeneProperties';
 import loadMarkerGenes from './loadMarkerGenes';
+import setGeneOrder from './setGeneOrder';
 
 export {
   loadGeneExpression,
   changeGeneSelection,
   loadPaginatedGeneProperties,
   loadMarkerGenes,
+  setGeneOrder,
```
Comment (ivababukova) [inline src/redux/actions/genes/index.js]: as I said elsewhere, genes store doesn't sound like the correct place to store the order of the genes in the Plots and Tables marker heatmap, consider moving it out.

File: src/redux/actions/genes/loadMarkerGenes.js [hunk 13]
```diff
@@ -37,6 +37,7 @@ const loadMarkerGenes = (experimentId, resolution) => async (dispatch, getState)
         experimentId,
         genes: order,
         data: markerGeneExpressions,
+        plotUuid,
```
Comment (ivababukova) [inline src/redux/actions/genes/loadMarkerGenes.js:40]: again, put plotType here

File: src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx (hunk 3 above)
Comment (StefanBabukov) [reply src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx]: I use `currentSelectedGenes` for when the loadexpression fails and we want to rerun it with the old genes which are not in redux yet.

File: src/redux/actions/genes/loadMarkerGenes.js (hunk 9 above)
Comment (StefanBabukov) [reply src/redux/actions/genes/loadMarkerGenes.js:8]: Yeah but if you look at `markerGenesLoaded.js` you can see that the genes are put under expression.views[plotUuid] thats where selectedGenes are stored. If we have another markerheatmap plot it will have different selectedGenes depending on what the user chose.

Comment (StefanBabukov) [reply src/redux/actions/genes/loadMarkerGenes.js:8]: or should this still be plotType?

**Comment by vickymorrison** [comment]
Metadata track bug - menu says Louvain track is shown but it does not appear on the plot.
Screenshot 2021-08-13 at 14 58 33

**Comment by vickymorrison** [comment]
The re-loading has a lag: I make a change and the heatmap disappears but then the OLD version reappears for a few seconds before the NEW version appears. Video example below of changing the number of genes per custer. This lag is 10 seconds.
https://user-images.githubusercontent.com/74908153/129368438-ccb657cd-acac-415c-bc7d-34f4165c6c6a.mov

**Comment by vickymorrison** [comment]
When scratchpad is selected in the menu I would expect the marker heatmap to update to show me markers for the custom cell sets. This does not happen. 
Screenshot 2021-08-13 at 15 02 01

Following a Slack chat this likely needs bioinformatics input - happy for you to merge this ticket and create a new one for this issue.

**Comment by StefanBabukov** [comment]
The re-loading has a lag: I make a change and the heatmap disappears but then the OLD version reappears for a few seconds before the NEW version appears. Video example below of changing the number of genes per custer. This lag is 10 seconds.
https://user-images.githubusercontent.com/74908153/129368438-ccb657cd-acac-415c-bc7d-34f4165c6c6a.mov
------------
I'm pretty sure this is how rendering vega plots works - shows the previous state of the plot and when its rendered shows the new one - we should work on improving the rendering time in a future ticket (downsampling possibly)

File: src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx (hunk 3 above)
Comment (ivababukova) [reply src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx]: the heatmap in plots and tables seems to be able to do that without the need for an extra state variable. Why can't we do this in here too?

File: src/__test__/components/plots-and-tables/markerHeatmap.test.jsx [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+  it('sorts genes properly', async () => {
+    await renderHeatmapPage();
+    await rtl.waitFor(() => expect(configUpdatedSpy).toHaveBeenCalled());
+
+    store.dispatch(loadGeneExpression(experimentId, ['gene0', 'gene1', 'gene3', 'gene2'], plotUuid));
+    await rtl.waitFor(() => expect(configUpdatedSpy).toHaveBeenCalledTimes(2));
+    expect(store.getState().componentConfig[plotUuid].config.selectedGenes).toEqual(['gene0', 'gene1', 'gene2', 'gene3']);
+  });
+
+  it('removing a gene keeps the sorted order', async () => {
```
Comment (ivababukova) [inline src/__test__/components/plots-and-tables/markerHeatmap.test.jsx]: also check that it is not re-sorting them after removing a gene

File: src/__test__/components/plots-and-tables/markerHeatmap.test.jsx [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+    const nGenesInput = rtl.getByTestId(geneSelection.parentElement, 'num-genes');
+    userEvent.type(nGenesInput, 10);
+
+    const runButton = rtl.getByText(geneSelection.parentElement, 'Run');
+    userEvent.click(runButton);
+
+    expect(loadMarkersSpy).toHaveBeenCalled();
+  });
+
+  it('sorts genes properly', async () => {
```
Comment (ivababukova) [inline src/__test__/components/plots-and-tables/markerHeatmap.test.jsx]: add tests for cases when you are adding or removing genes, but sort is not necessary.

Comment (StefanBabukov) [reply src/__test__/components/plots-and-tables/markerHeatmap.test.jsx]: This test is for that when you add a gene - checks if the genes are sorted properly with the newly added gene

File: src/__test__/components/plots-and-tables/markerHeatmap.test.jsx (hunk 14 above)
Comment (StefanBabukov) [reply src/__test__/components/plots-and-tables/markerHeatmap.test.jsx]: Also added that test too look at the changes

File: src/__test__/components/plots-and-tables/markerHeatmap.test.jsx [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+    <Provider store={store}>
+      <MarkerHeatmap
+        experimentId={experimentId}
+      />
+    </Provider>,
+  );
+  await rtl.waitFor(() => expect(loadConfigSpy).toHaveBeenCalledTimes(1));
+};
+
+describe('Marker heatmap plot', () => {
```
Comment (ivababukova) [inline src/__test__/components/plots-and-tables/markerHeatmap.test.jsx:187]: that's some good addition of tests, but it is not sufficient. Add tests to check for what happens if the redux store is empty, if parts of the redux store are empty, if after requesting expression an error is returned and so on. Try to think what can possibly go wrong in this file and add a test for it.

**Comment by vickymorrison** [comment]
> ## The re-loading has a lag: I make a change and the heatmap disappears but then the OLD version reappears for a few seconds before the NEW version appears. Video example below of changing the number of genes per custer. This lag is 10 seconds.
> https://user-images.githubusercontent.com/74908153/129368438-ccb657cd-acac-415c-bc7d-34f4165c6c6a.mov
> 
> I'm pretty sure this is how rendering vega plots works - shows the previous state of the plot and when its rendered shows the new one - we should work on improving the rendering time in a future ticket (downsampling possibly)

But the issue is that the heatmap disappears while it is being rendered (with "we're getting your data" message), but then the OLD one appears for 5-10secs before the NEW one eventually loads. It would be better if there was simply a longer time of "we're getting your data" as it is confusing to the user that the OLD version reappears.

File: src/__test__/redux/reducers/genes/markerGenesLoaded.test.js [hunk 17]
```diff
@@ -25,6 +25,7 @@ describe('markerGenesLoaded', () => {
       payload: {
         genes,
         data,
+        plotUuid: 'interactiveHeatmap',
```
Comment (ivababukova) [inline src/__test__/redux/reducers/genes/markerGenesLoaded.test.js:28]: change to `plotType`

Comment (ivababukova) [reply src/__test__/redux/reducers/genes/markerGenesLoaded.test.js:28]: or actually, judging from file `src/components/data-exploration/heatmap/HeatmapPlot.jsx` it should be `componentType`

File: src/__test__/components/plots-and-tables/markerHeatmap.test.jsx (hunk 16 above)
Comment (ivababukova) [reply src/__test__/components/plots-and-tables/markerHeatmap.test.jsx:187]: There are also lots of small functions that are not covered like `getCellClusterFromCellId` -- we need them tested too. Looking at the function `getCellClusterFromCellId`, looks like it will crash if `properties` is undefined or if there is no `clusters` or if `clusters` are empty. 
We have to have guards against all these things and make sure that the code works not only if the user uses it exactly as they are intended to, but when they try to do something else. 
Throughout all this code, try to apply a mindset towards programming for failure -- go through each line and think: "how can possibly this line break?" and then add handling for that and a unit test. Right now, it feels like this is programmed for the golden path use case. I am concerned that if user attempts to do something slightly different, the heatmap will crash.

File: src/__test__/components/plots-and-tables/markerHeatmap.test.jsx [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+    const nGenesInput = rtl.getByTestId(geneSelection.parentElement, 'num-genes');
+    userEvent.type(nGenesInput, 10);
+
+    const runButton = rtl.getByText(geneSelection.parentElement, 'Run');
+    userEvent.click(runButton);
+
+    expect(loadMarkersSpy).toHaveBeenCalled();
+  });
+
+  it('sorts genes properly when adding a gene', async () => {
```
Comment (ivababukova) [inline src/__test__/components/plots-and-tables/markerHeatmap.test.jsx:223]: also missing tests for the heatmap states: that when we are loading a gene, the heatmap is in loading state. When there's been an error in loading a gene, the heatmap is in error state.

Comment (StefanBabukov) [reply src/__test__/components/plots-and-tables/markerHeatmap.test.jsx:223]: Do we  not have tests for this - loadGeneExpression do i  need to test for that here again?

File: src/components/plots/helpers/populateHeatmapData.js [hunk 19]
```diff
@@ -121,9 +142,11 @@ const populateHeatmapData = (
     // we want to avoid displaying elements which are not in a louvain cluster
     // so initially consider as enabled only cells in louvain clusters
     // See: https://biomage.atlassian.net/browse/BIOMAGE-809
-    const louvainClusters = hierarchy.find((clusters) => clusters.key === 'louvain');
-
-    const cellIsInLouvainCluster = getCellsSetInGroup(louvainClusters);
+    const selectedCellSet = heatmapSettings?.selectedCellSet ? heatmapSettings.selectedCellSet : 'louvain';
```
Comment (ivababukova) [inline src/components/plots/helpers/populateHeatmapData.js:145]: I don't understand the addition of line 145, can you explain? Also, is it possible to remove the hardcoding of `louvain` in here? In the very hear future, we will want to support other types of clustering, not only louvain, and this line will be a problem

File: src/components/plots/styling/PlotStyling.jsx [hunk 20]
```diff
@@ -19,8 +19,6 @@ import ColourInversion from './ColourInversion';
 import ExpressionValuesType from './ExpressionValuesType';
 import ExpressionValuesCapping from './ExpressionValuesCapping';
 
-import MetadataTracksDisplayEditor from './MetadataTracksDisplayEditor';
```
Comment (ivababukova) [inline src/components/plots/styling/PlotStyling.jsx:22]: why is this no longer used?

File: src/components/plots/helpers/populateHeatmapData.js (hunk 19 above)
Comment (StefanBabukov) [reply src/components/plots/helpers/populateHeatmapData.js:145]: This is for the 'Select data' menu, when we don't have that option in the config - use 'louvain', don't know how can that be not hardcoded at the moment.

File: src/components/plots/styling/PlotStyling.jsx (hunk 20 above)
Comment (StefanBabukov) [reply src/components/plots/styling/PlotStyling.jsx:22]: Duplicate component - look at `HeatmapMetadataTrackSettings`

File: src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx [hunk 21]
```diff
@@ -0,0 +1,307 @@
...
+    if (louvainClustersResolution && config) {
+      dispatch(loadMarkerGenes(experimentId, louvainClustersResolution, plotUuid, config.numGenes));
+    }
+  }, [louvainClustersResolution, hierarchy, config?.numGenes]);
+  const sortGenes = (newGenes) => {
+    const clusters = hierarchy.find((cluster) => cluster.key === config.selectedCellSet).children;
+
+    const getCellIdsForCluster = (clusterId) => properties[clusterId].cellIds;
+
+    const getAverageExpressionForGene = (gene, currentCellIds) => {
```
Comment (ivababukova) [inline src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx:70]: this function is not unit tested either

Comment (StefanBabukov) [reply src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx:70]: It is tested - its part of the sorting and sorting is tested and working.

File: src/__test__/components/plots-and-tables/markerHeatmap.test.jsx (hunk 16 above)
Comment (StefanBabukov) [reply src/__test__/components/plots-and-tables/markerHeatmap.test.jsx:187]: These functions are tested - they are used when sorting, which is tested. What can crash it, if there are no cellsets  - we load them - there wont be an option to sort anything if they are not there.

File: src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx [hunk 22]
```diff
@@ -0,0 +1,307 @@
...
+      const expressionValues = expressionData.data[gene].rawExpression.expression;
+      let totalValue = 0;
+      currentCellIds.forEach((cellId) => {
+        totalValue += expressionValues[cellId];
+      });
+      return totalValue / currentCellIds.size;
+    };
+
+    const getClusterForGene = (gene) => {
+      let maxAverageExpression = [0];
```
Comment (ivababukova) [inline src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx]: I think that the default value of `maxAverageExpression` as a list with 0 is confusing -- what you technically have is a list with 2 items, where the first index shows the value of the max expression and the second shows the index of the cluster where this value occurs. 
The code will be cleaner if you either:
- use 2 variables for this: `maxAvgExpression`, just a number denoting max expression; and `maxExpressionClusterIndex` denoting the cluster index when it occured. Then in line 90 you just return `maxExpressionClusterIndex`.
- use 1 object for this, called `maxAvgExpression` that looks like something like this:
```
let maxAvgExpression = {"expression": 0, "clusterIndex": -1}
```
then on line 90 you return maxAvgExpression["clusterIndex"]

Comment (ivababukova) [reply src/pages/experiments/[experimentId]/plots-and-tables/marker-heatmap/index.jsx]: another thing: are you sure that it is not possible to have negative expression? If negative expression is possible, then setting default maxExpression to 0 is incorrect.

File: src/__test__/components/plots-and-tables/markerHeatmap.test.jsx (hunk 18 above)
Comment (ivababukova) [reply src/__test__/components/plots-and-tables/markerHeatmap.test.jsx:223]: in loadGeneExpression we test the store, in here you would test if the component correctly responds to the updates of the store. So you would check that `Skeleton` gets rendered when the relevant parts of the redux store are in loading state. What you can do is to mock the store, settings some of the contents to be loading: true (like cellSets or geneExpression for this plotUuid) and then assert that instead of MarkerHeatmap, we render `Skeleton`.

File: src/__test__/components/plots-and-tables/markerHeatmap.test.jsx [hunk 23]
```diff
@@ -0,0 +1,268 @@
...
+
+    await rtl.waitFor(() => expect(loadMarkersSpy).toHaveBeenCalled());
+  });
+
+  it('sorts genes properly when adding a gene', async () => {
+    await renderHeatmapPage(defaultStore);
+    await rtl.waitFor(() => expect(configUpdatedSpy).toHaveBeenCalled());
+    store.dispatch(loadGeneExpression(experimentId, ['gene0', 'gene1', 'gene3', 'gene2'], plotUuid));
+    await rtl.waitFor(() => expect(configUpdatedSpy).toHaveBeenCalledTimes(2));
+    expect(store.getState().componentConfig[plotUuid].config.selectedGenes).toEqual(['gene0', 'gene1', 'gene2', 'gene3']);
```
Comment (ivababukova) [inline src/__test__/components/plots-and-tables/markerHeatmap.test.jsx:228]: much, much better now

File: src/__test__/components/plots-and-tables/markerHeatmap.test.jsx [hunk 24]
```diff
@@ -0,0 +1,268 @@
...
+    expect(loadMarkersSpy).toHaveBeenCalledTimes(1);
+  });
+
+  it('loads cellsets if not available', async () => {
+    const newStore = { ...defaultStore, cellSets: { loading: true } };
+    await renderHeatmapPage(newStore);
+    await rtl.waitFor(() => expect(cellSetsLoadedSpy).toHaveBeenCalled());
+  });
+
+  it('loads processing settings if louvainresolution is not available', async () => {
```
Comment (ivababukova) [inline src/__test__/components/plots-and-tables/markerHeatmap.test.jsx:256]: nice, this is what I meant on the call. Also add a test with `newStore` equal to what is its initial value, then try to load the marker heatmap and assert that the heatmap goes into loading state (Skeleton is displayed) and that dispatches for the relevant actions have been triggered.

File: src/components/plots/helpers/populateHeatmapData.js (hunk 19 above)
Comment (ivababukova) [reply src/components/plots/helpers/populateHeatmapData.js:145]: fair enough, let's worry about it later then

File: src/components/plots/helpers/populateHeatmapData.js [hunk 25]
```diff
@@ -238,7 +260,7 @@ const populateHeatmapData = (
 
   data.trackColorData = trackData.map((datum) => datum.trackColorData).flat();
   data.trackGroupData = trackData.map((datum) => datum.groupData).flat();
-
+  data.clusterSeparationLines = trackData[0].clusterSeparationLines;
```
Comment (ivababukova) [inline src/components/plots/helpers/populateHeatmapData.js:263]: is it possible that `trackData` will be empty? Because then `trackData[0]` will error out

File: src/components/plots/styling/heatmap/HeatmapControls.jsx [hunk 26]
```diff
@@ -26,18 +60,78 @@ const HeatmapControls = (props) => {
...
         </Space>
       </Panel>
-      <Panel header='Metadata tracks' key='11'>
+      {markerHeatmap && (
+        <>
+          <Panel header='Select Data' key='selectData'>
+            <p>Select the cell sets to show markers for:</p>
+            <Select
+              value={{
+                key: firstLetterUppercase(config.selectedCellSet),
```
Comment (ivababukova) [inline src/components/plots/styling/heatmap/HeatmapControls.jsx:111]: why aren't you using the key of the selectedCellSet? That should be unique, whereas we can have two or more cell sets with the same name

File: src/utils/plots/getSelectOptions.js [hunk 27]
```diff
@@ -0,0 +1,18 @@
+// this returns a formatted array for the select element
+
+const getSelectOptions = (options) => {
```
Comment (ivababukova) [inline src/utils/plots/getSelectOptions.js:3]: look at whether we also use this in other files and also use it there

File: src/pages/experiments/[experimentId]/data-exploration/index.jsx [hunk 28]
```diff
@@ -60,7 +60,7 @@ const ExplorationViewPage = ({
     Heatmap: {
       toolbarControls: (
         <>
-          <HeatmapSettings componentType={COMPONENT_TYPE} key='heatmap-settings' />
+          <HeatmapSettings componentType={COMPONENT_UUID} key='heatmap-settings' />
```
Comment (ivababukova) [inline src/pages/experiments/[experimentId]/data-exploration/index.jsx]: I know I asked you to make everything the same type and this is why you changed it to component_uuid, but I didn't realise that now this is going to be confusing here: `componentType={COMPONENT_UUID}`. Sounds like it was maybe better to use component type for data exploration and component uuid for plots and tables. Can you change that back? Sorry!

**Review by ivababukova — COMMENTED** [review]
some very small things left here in terms of code. Did you look at the loading displyaing previous state briefly issue reported by Vicky?

File: src/components/plots/helpers/populateHeatmapData.js (hunk 25 above)
Comment (StefanBabukov) [reply src/components/plots/helpers/populateHeatmapData.js:263]: no, it will return something from generateTrackData

**Review by ivababukova — APPROVED** [review]
